include/blmc_robots/real_finger.hpp

```diff
@@ -15,62 +15,133 @@
+    double torque_ratio;  // TODO better used fixed torque
+    //! @brief P-gain for the position controller.
+    double control_gain_kp;
+    //! @brief D-gain for the position controller.
+    double control_gain_kd;
+    //! @brief Tolerance for reaching the starting position.
+    double position_tolerance_rad;
+    //! @brief Timeout for reaching the starting position.
+    double move_timeout;
+};
+
+
+// TODO add "BLMC" in the name?
+/**
+ * @brief Base class for simple n-joint BLMC robots.
+ *
+ * This is a generic base class to easily implement drivers for simple BLMC
+ * robots that consist of a chain of joints.
+ *
+ * @tparam N_JOINTS Number of joints.
+ * @tparam N_MOTOR_BOARDS Number of motor control boards that are used.
+ */
+template <size_t N_JOINTS, size_t N_MOTOR_BOARDS>
+class NJointBlmcRobotDriver
+    : public robot_interfaces::RobotDriver<
+          typename robot_interfaces::NJointRobotTypes<N_JOINTS>::Action,
+          typename robot_interfaces::NJointRobotTypes<N_JOINTS>::Observation>
 {
 public:
-    typedef robot_interfaces::finger::Action Action;
-    typedef robot_interfaces::finger::Observation Observation;
-    typedef robot_interfaces::finger::Vector Vector;
-    typedef std::array<std::shared_ptr<blmc_drivers::MotorInterface>, 3> Motors;
-    typedef std::array<std::shared_ptr<blmc_drivers::CanBusMotorBoard>, 2>
+    typedef
+        typename robot_interfaces::NJointRobotTypes<N_JOINTS>::Action Action;
+    typedef typename robot_interfaces::NJointRobotTypes<N_JOINTS>::Observation
+        Observation;
+    typedef
+        typename robot_interfaces::NJointRobotTypes<N_JOINTS>::Vector Vector;
+    typedef std::array<std::shared_ptr<blmc_drivers::MotorInterface>, N_JOINTS>
+        Motors;
+    typedef std::array<std::shared_ptr<blmc_drivers::CanBusMotorBoard>,
+                       N_MOTOR_BOARDS>
         MotorBoards;
 
-    RealFingerDriver(const std::string &can_0, const std::string &can_1)
-        : RealFingerDriver(create_motor_boards(can_0, can_1))
+    NJointBlmcRobotDriver(const std::array<std::string, N_MOTOR_BOARDS> &can_ports,
+                      const MotorParameters &motor_parameters,
+                      const double max_action_duration_s,
+                      const double max_inter_action_duration_s)
+        : NJointBlmcRobotDriver(create_motor_boards(can_ports),
+                            motor_parameters,
+                            max_action_duration_s,
+                            max_inter_action_duration_s)
     {
     }
 
-    RealFingerDriver(const MotorBoards &motor_boards)
-        : RealFingerDriver(create_motors(motor_boards))
+    NJointBlmcRobotDriver(const MotorBoards &motor_boards,
+                      const MotorParameters &motor_parameters,
+                      const double max_action_duration_s,
+                      const double max_inter_action_duration_s)
+        : NJointBlmcRobotDriver(create_motors(motor_boards),
```

Comment:
Right, I missed that part. I think passing some motor-to-board mapping and construct `Motors` from this would be unnecessarily complicated. How about in `NJointBlmcRobotDriver` we only keep a constructor that already takes the `Motors` array and move the `create_motors` logic to the derived classes?

Reply:
done

Reply:
thanks! 

the concern i have now is that it is not super logical, because we have a create_motor_boards function, but not a create_motors function, so i think it would make more sense to either have both inside of both outside of the class.